plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -154,7 +105,11 @@ fullyApplyAndBetaReduce info args0 = do
               (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
               acc
           go acc' arity' args'
-        _ -> pure Nothing
+        -- term/type argument mismatch, don't inline
+        (TermParam _:_, TypeAppContext{}) -> pure Nothing
+        (TypeParam _:_, TermAppContext{}) -> pure Nothing
+        -- no more arguments to apply, just apply what we have
+        (_, AppContextEnd) -> pure $ Just acc
```

Comment:
You can't just return `acc` here - there are remaining lambdas.

Consider `f x y` where `f = \a b c d -> a + b + c + d`.
You need to return `\c d -> x + y + c + d`.
But `acc` is `x + y + c + d`.